### ISSUE TITLE
The "React" way of setting up user sign in and protected routes

### ISSUE THREAD

**Issue by ylorenzana** [kentcdodds/ama#431]
I'm working on an example client that can register/sign in users. The client will consume an API that uses bearer token based authentication (basically server-side sessions). I have the user registration and sign in figured out, I just find myself stuck thinking about the "right" way of implementing a protected route for production in React. The way I have it planned out, the /home page of the app will be protected, requiring the user to be authenticated to access that route. So here is why I'm confused:
* Should I have protected routes render conditionally, based on making a _request every time_ to the /api/users/me endpoint of my back-end API (this endpoint goes through authentication middleware, takes in session cookie and returns the email associated to the session if successful) __before loading__?
  - If this were the case, how would this be implemented? Using one of the component lifecycle methods to send the http request before loading, and setting state that keeps track of whether the user is authenticated, then using conditional rendering (based on the authentication state) to either render home page or redirect to login page?
* Should I use localStorage instead? After logging in or registering succesfully, I could set something in localStorage like ('email', {usersEmail}). Then instead of having to send a request to the server every time a user wants to load the home page, the home component instead checks if there is a value for email in localStorage, and if there is render, the home page, otherwise redirect to login. In this case, the home page wouldn't be truly "protected", as the session cookie could expire, yet the value in localStorage would still be set. This doesn't sound secure either. Keep in mind storing the session token in the client is __not__ an option.

At this point, I don't really know what approach to take. This is a lot easier to implement using client-side tokens like JWTs, but again, I want to learn how to implement this using server-side sessions. I figured this is a very common problem, and hopefully there is *some* consensus on what the _right way_ of doing this in React. 

Any help at all would be greatly appreciated. Thanks in advance!

**Comment by hedgerh** [comment]
I'd go with the first option.  You don't want them to be able to visit protected routes when their session is expired.  They won't be able to get the data to load any of the protected pages or do anything that requires authentication.

You can redirect or turn set the auth state to false when you get 403s from the api.  Not sure what you're using for routing, but I think react-router has an auth example in their repo.

Also, I recommend joining the Reactiflux community if you haven't already.  It's a good place to get answers for questions like these.

**Comment by kentcdodds** [comment]
Hi! I'm afraid I don't have time to give a super solid answer to all my open questions right now, so I just went through and answered all my AMA questions in the form of a video: https://www.youtube.com/watch?v=y23g8L8UtoQ&list=PLV5CVI1eNcJgCrPH_e6d57KRUTiDZgs0u

A few things that folks can do to help make things better for others reading the questions and answers:

1. Link to when I start answering the question. Add `&t=xxmxxs` to the end of the URL like this: https://www.youtube.com/watch?v=y23g8L8UtoQ&list=PLV5CVI1eNcJgCrPH_e6d57KRUTiDZgs0u&t=2m5s
2. Comment with links I referenced
3. Summarize what I talk about

Thanks!
